calculator/calculator.py

```diff
@@ -1,15 +1,123 @@
 import re
+import math
+from typing import Union
+from statistics import mean
 
 
-class Calculator:
+_NUMBER_REGEX = r'\d+(?:\.\d+)?'
+_NUMBER_NEGATIVE_OPT_REGEX = f'-?{_NUMBER_REGEX}'
+_LEFT_PARENTHESIS_REGEX = r'(?P<LEFT_PARENTHESIS>\()'
+_RIGHT_PARENTHESIS_REGEX = r'(?P<RIGHT_PARENTHESIS>\))'
+_FACTORIAL_REGEX = r'(?P<FACTORIAL>!)'
+_NEGATE_REGEX = r'(?P<NEGATE>~)'
+_AVERAGE_REGEX = r'(?P<AVERAGE>@)'
+_MIN_REGEX = r'(?P<MIN>\&)'
+_MAX_REGEX = r'(?P<MAX>\$)'
+_POWER_REGEX = r'(?P<POWER>\^)'
+_MULTIPLICATION_REGEX = r'(?P<MULTIPLICATION>\*)'
+_DIVISION_REGEX = r'(?P<DIVISION>/)'
+_ADDITION_REGEX = r'(?P<ADDITION>\+)'
+_SUBTRACTION_REGEX = r'(?P<SUBTRACTION>-)'
 
-    @classmethod
-    def calculate(cls, expression=str):
-        if type(expression) is not str:
-            raise TypeError("Expression input should be of str type.")
 
-        # TODO
+def calculate(expression: str) -> Union[int, float]:
+    if type(expression) is not str:
+        raise TypeError("Expression input should be of str type.")
 
+    return _calculate(expression.replace(" ", ""))
 
+def _calculate(expression: str) -> Union[int, float]:
+    if _is_end_of_calculation(expression):
+        return expression
 
+    for regex, handler in _TOKEN_SPECIFICATION.items():
+        tmp_result = _handle(expression, regex, handler)
+        if tmp_result is not None:
+            return _calculate(tmp_result)
 
+    raise ValueError(f'Expression: {expression} is not legal!')
+
+def _add_operation_variables(operation: str, is_single_variable: bool = False, is_negative_legal: bool = False) -> str:
+    number_regex = _NUMBER_NEGATIVE_OPT_REGEX if is_negative_legal else _NUMBER_REGEX
+    return f'(?P<NUMBER>{number_regex}){operation}' if is_single_variable else f'(?P<FIRST_NUMBER>{number_regex})' \
+                                                                               f'{operation}' \
+                                                                               f'(?P<SECOND_NUMBER>{number_regex})'
+
+def _is_end_of_calculation(expression: str) -> bool:
+    return re.fullmatch(f'{_NUMBER_NEGATIVE_OPT_REGEX}', expression) is not None
+
+def _handle(expression: str, regex: str, handler) -> str:
+    token = re.search(regex, expression)
+    if token:
+        return re.sub(token.re, handler(token), expression, 1)
+
+def _handle_two_variable_operation(func):
+    def _inner(token):
+        return str(func(float(token.group("FIRST_NUMBER")), float(token.group("SECOND_NUMBER"))))
+    return _inner
+
+def _handle_single_variable_operation(func):
+    def _inner(token):
+        return str(func(float(token.group("NUMBER"))))
```

Comment:
Same methods except the arguments.
Read about groups() method and make it one generic function.